### ISSUE TITLE
Ensure database unique relationships are well defined

### ISSUE THREAD

**Issue by sfount** [third-culture-software/bhima#347]
This issue tracks re-factoring the SQL schema keys. 

We currently have many entities in BHIMA that are uniquely identified by an UUID. However we very rarely enforce other unique constraints on tables. 

There are many examples of entities that should be considered unique on attributes (name, account, etc.), ensuring these are well defined now will save us from many errors in the future. 

For example - `debtor_group` 

Right now the only constraint is 

``` sql
PRIMARY KEY(`uuid`)
```

This means that all of the following are valid:

| uuid | name | account_id |
| --- | --- | --- |
| fe543 | Example Debtor | 250 |
| bd394 | Example Debtor | 250 |
| se293 | Example Debtor | 250 |

We can visually see that all of these debtor groups are identical however there is no rule specifying that in the database. This could lead to allowing users to record bad data. 

**Proposal:**
- Use the `CONSTRAINT ... UNIQUE` keyword in SQL to formalise definition for all tables in the BHIMA database. 

For example 

``` sql
CONSTRAINT debtorName UNIQUE (name, account_id)
```
- Ensure `description` is `NOT NULL` everywhere this should be depended on. 
- Extend the current error handling middle-ware to ensure the now covered corner cases are reported correctly.

**Comment by jniles** [comment]
I'm happy with this proposal - it will be a huge boost to our data integrity at minimal (if any!) inconvenience to our users.  If we get a report that this is a poor design choice for some table, we can always roll back the change for that specific scenario.

I would also recommend that whoever does this ensure that `description` is `NOT NULL` everywhere we depend on it.  For example, we will always display the `sale.description`, `cash.description`, etc, labels.  These should be `NOT NULL` so that we don't accidentally create meaningless information.

Finally, the developer who tackles this issue should make sure that intelligent error codes are returned for `UNIQUE` violations.

**Comment by sfount** [comment]
@jniles Thanks for the analysis, I've updated the description to factor in your comment. 

@DedrickEnc This requires a good understanding of the database and a proposal for how the unique relationships should look and where they should be - are you able to work on this next?

**Comment by DedrickEnc** [comment]
Thanks

**Comment by jniles** [comment]
Closed in #513.
